src/matrix_ops.cpp

```diff
@@ -0,0 +1,189 @@
+/*
+ *
+ * matrix_ops.cpp
+ * Distance matrix transformations
+ *
+ */
+#include <vector>
+#include <thread>
+#include <cstdint>
+#include <cstddef>
+#include <cmath>
+#include <string>
+
+#include "api.hpp"
+
+// Prototypes for cppying functions run in threads
+void square_block(const Eigen::VectorXf& longDists,
+                  SquareMatrix& squareMatrix,
+                  const size_t n_samples,
+                  const size_t start,
+                  const size_t offset,
+                  const size_t max_elems);
+
+void rectangle_block(const Eigen::VectorXf& longDists,
+                  SquareMatrix& squareMatrix,
+                  const size_t nrrSamples,
+                  const size_t nqqSamples,
+                  const size_t start,
+                  const size_t max_elems);
+
+template<class T>
+size_t rows_to_samples(const T& longMat) {
+    return 0.5*(1 + sqrt(1 + 8*(longMat.rows())));
+}
+
+long calc_row_idx(const long long k, const long n) {
+	return n - 2 - floor(sqrt((double)(-8*k + 4*n*(n-1)-7))/2 - 0.5);
+}
+
+long calc_col_idx(const long long k, const long i, const long n) {
+	return k + i + 1 - n*(n-1)/2 + (n-i)*((n-i)-1)/2;
+}
+
+long long square_to_condensed(long i, long j, long n) {
+    assert(i > j);
+	return (n*j - ((j*(j+1)) >> 1) + i - 1 - j);
+}
+
+std::tuple<size_t, unsigned int, unsigned int> 
+   jobs_per_thread(const size_t num_jobs, 
+                   const unsigned int num_threads) {
+    size_t used_threads = num_threads;
+    if (num_jobs < used_threads) {
+        used_threads = num_jobs; 
+    }
+    size_t calc_per_thread = (size_t)num_jobs / used_threads;
+    unsigned int num_big_threads = num_jobs % used_threads;
+
+    return(std::make_tuple(calc_per_thread, used_threads, num_big_threads)); 
+}
+
+SquareMatrix long_to_square(const Eigen::VectorXf& rrDists, 
+                            const Eigen::VectorXf& qrDists,
+                            const Eigen::VectorXf& qqDists,
+                            unsigned int num_threads) {
+    // Set up square matrix to move values into
+    size_t nrrSamples = rows_to_samples(rrDists);
+    size_t nqqSamples = 0;
+    if (qrDists.size() > 0) {
```

Comment:
I don't know if there's ever a case in this code for having qr distances but not qq distances, e.g. when querying a database without updating it, and so whether it would be worth separating qr and qq distance calculations. But may well not be understanding how this is (or is intended to be) deployed.

Reply:
The reason for using query vs ref as the test here is that if you are adding a single query the qq length will be zero, but qr will always be > 0. I think the returned matrix is always expected to be square here, but I might be wrong. We'll see if it runs in to any problems when used in PopPUNK

Reply:
Hadn't thought about addition of one query, makes complete sense. I suppose the matrix will always be square (zero filled?) - in PopPUNK, there is the 'queryQuery' condition that dictates whether query-query distances are to be added - is there a risk the matrix could be left the right shape, but with QQ distances artificially left as 0 if no automatic update of the database?

Reply:
My idea was to do this somewhat in the python and leave the C++ general. So there are two exported functions to python, `longToSquare()` which just takes ref-ref distances, and `longToSquareMulti()` which takes rr, rq and qq. So there might need to be a bit of selection in the PopPUNK code on which of these to use, but the former should cope when `queryQuery = False`